src/Sulu/Bundle/PageBundle/EventListener/PageRemoveSubscriber.php

```diff
@@ -0,0 +1,191 @@
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\PageBundle\EventListener;
+
+use PHPCR\Query\QueryInterface;
+use PHPCR\Query\RowInterface;
+use Sulu\Bundle\PageBundle\Document\BasePageDocument;
+use Sulu\Bundle\PageBundle\Document\PageDocument;
+use Sulu\Bundle\SecurityBundle\AccessControl\AccessControlQueryEnhancer;
+use Sulu\Component\Content\Document\Behavior\SecurityBehavior;
+use Sulu\Component\DocumentManager\Event\ConfigureOptionsEvent;
+use Sulu\Component\DocumentManager\Event\RemoveEvent;
+use Sulu\Component\DocumentManager\Events;
+use Sulu\Component\PHPCR\SessionManager\SessionManagerInterface;
+use Sulu\Component\Rest\Exception\DependantResourcesFoundException;
+use Sulu\Component\Rest\Exception\InsufficientDescendantPermissionsException;
+use Sulu\Component\Security\Authentication\UserInterface;
+use Sulu\Component\Security\Authorization\PermissionTypes;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Security\Core\Security;
+
+class PageRemoveSubscriber implements EventSubscriberInterface
+{
+    const FORCE_REMOVE_CHILDREN_OPTION = 'force_remove_children';
+
+    /**
+     * @var SessionManagerInterface
+     */
+    private $sessionManager;
+
+    /**
+     * @var AccessControlQueryEnhancer
+     */
+    private $accessControlQueryEnhancer;
+
+    /**
+     * @var Security|null
+     */
+    private $security;
+
+    /**
+     * @var array<string, int>
+     */
+    private $permissions;
+
+    /**
+     * @param array<string, int> $permissions
+     */
+    public function __construct(
+        SessionManagerInterface $sessionManager,
+        AccessControlQueryEnhancer $accessControlQueryEnhancer,
+        ?Security $security,
+        array $permissions
+    ) {
+        $this->sessionManager = $sessionManager;
+        $this->accessControlQueryEnhancer = $accessControlQueryEnhancer;
+        $this->security = $security;
+        $this->permissions = $permissions;
+    }
+
+    /**
+     * @return array<string, mixed>
+     */
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            Events::REMOVE => ['handleRemove', 4096],
+            Events::CONFIGURE_OPTIONS => ['handleConfigureOptions'],
+        ];
+    }
+
+    /**
+     * @param array<array{id: string, resourceKey: string, depth: int}> $resources
+     *
+     * @return array<int, array<array{id: string, resourceKey: string}>>
+     */
+    private function groupResourcesByDepth(array $resources): array
+    {
+        $grouped = [];
+
+        foreach ($resources as $resource) {
+            $depth = $resource['depth'];
+            unset($resource['depth']);
+
+            if (!isset($grouped[$depth])) {
+                $grouped[$depth] = [];
+            }
+
+            $grouped[$depth][] = $resource;
+        }
+
+        \krsort($grouped);
+
+        return \array_values($grouped);
+    }
+
+    public function handleRemove(RemoveEvent $event): void
```

Comment:
this should only be `validateIfReadyToRemove`? Or something? `handleRemove` sounds like the actual remove